staging/src/k8s.io/apiserver/pkg/endpoints/discovery/root/discovery.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package root
+
+import (
+	"net/http"
+
+	restful "github.com/emicklei/go-restful/v3"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/endpoints/discovery/aggregated"
+	"k8s.io/apiserver/pkg/endpoints/handlers/negotiation"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+)
+
+const APIGroupPrefix = "/apis"
+
+type RootManager interface {
+	WebService() *restful.WebService
+}
+
+// RootAPIsHandler creates a webservice serving /apis
+type RootAPIsHandler struct {
+	serializer                      runtime.NegotiatedSerializer
+	discoveryGroupManager           http.Handler
+	aggregatedDiscoveryGroupManager http.Handler
+}
+
+func NewRootAPIsHandler(serializer runtime.NegotiatedSerializer, discoveryGroupManager, aggregatedDiscoveryGroupManager http.Handler) *RootAPIsHandler {
+	return &RootAPIsHandler{
+		serializer:                      serializer,
+		discoveryGroupManager:           discoveryGroupManager,
+		aggregatedDiscoveryGroupManager: aggregatedDiscoveryGroupManager,
+	}
+}
+
+func (s *RootAPIsHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AggregatedDiscoveryEndpoint) {
+		mediaType, _ := negotiation.NegotiateMediaTypeOptions(req.Header.Get("Accept"), s.serializer.SupportedMediaTypes(), aggregated.DiscoveryEndpointRestrictions)
+		// mediaType.Convert looks at the request accept headers and is used to control whether the discovery document will be aggregated.
+		switch aggregated.IsAggregatedDiscoveryGVK(mediaType.Convert) {
+		case true:
+			s.aggregatedDiscoveryGroupManager.ServeHTTP(resp, req)
+			return
+		default:
+			s.discoveryGroupManager.ServeHTTP(resp, req)
+			return
+		}
+	}
+	s.discoveryGroupManager.ServeHTTP(resp, req)
+}
+
+func (s *RootAPIsHandler) restfulHandle(req *restful.Request, resp *restful.Response) {
+	s.ServeHTTP(resp.ResponseWriter, req.Request)
+}
+
+// WebService returns a webservice serving api discovery.
```

Comment:
Do we need a TODO here to mention the other issue you created relative to content-type in the openapi?

Reply:
I think https://github.com/kubernetes/kubernetes/issues/113028 is strictly an OpenAPI issue and covers more than just this. I'm also not sure if this is even the best place to put a TODO (within negotiation.MediaTypesForSerializer is a more likely place) that I don't think we should add it here.